src/Sparrow/Json/JsonContextPoolBase.cs

```diff
@@ -1,24 +1,30 @@
 ﻿using System;
 using System.Collections.Concurrent;
+using System.Collections.Generic;
 using System.Threading;
 using Sparrow.LowMemory;
+using Sparrow.Platform;
 using Sparrow.Threading;
 using Sparrow.Utils;
 
 namespace Sparrow.Json
 {
-    public abstract class JsonContextPoolBase<T> : ILowMemoryHandler, IDisposable 
+    public abstract class JsonContextPoolBase<T> : ILowMemoryHandler, IDisposable
         where T : JsonOperationContext
     {
+        private readonly object _locker = new object();
+
         private long _generation;
         private bool _disposed;
 
         protected SharedMultipleUseFlag LowMemoryFlag = new SharedMultipleUseFlag();
         private readonly CancellationTokenSource _cts = new CancellationTokenSource();
         private readonly long _maxContextSizeToKeepInBytes;
+        private readonly long _maxNumberOfContextsToKeepInGlobalStack;
 
         private readonly T[][] _perCoreContexts;
-        private readonly ConcurrentQueue<T> _globalQueue = new ConcurrentQueue<T>();
+        private CountingConcurrentStack<T> _previousGlobalStack;
```

Comment:
Why the change from queue to stack?

Reply:
If we have a queue with 4000 items in it already and we are requesting let's say 20 contexts per second (synchronously even - no concurrency). It means that in 1min we will take 1200 different contexts from the pool and refresh their 'last used date', so the cleaner will not trigger, but by using stack, we will use only 1 context from the stack, so others can be safely removed by the cleaner.

Reply:
IIRC, there was a reasoning behind this.
The idea is that we aren't going to end up with so many items in the global pool, because that is quite strange to have.
And when you pull from the global pool it goes to the per core one, no?
That would avoid this issue.

The auto cleanup of old instance really helps here.